src/include/interval_set.h

```diff
@@ -209,7 +209,7 @@ class interval_set {
   };
 
   interval_set() = default;
-  interval_set(Map&& other) {
+  interval_set(Map& other) {
```

Comment:
Instead of a straight revert, let's fix the bug by changing this to:

```suggestion
  interval_set(Map other) {
```

So it's a regular copy. It's up to the caller to allow the copy or pass `std::move(m)` to avoid the copy.

Reply:
> So it's a regular copy. It's up to the caller to allow the copy or pass `std::move(m)` to avoid the copy.

Since `interval_set` is a vocabulary type, it would be nice to provide both copy and move variants for maximum efficiency.

Reply:
A constructor like `interval_set(Map other)` captures both copy and move:

* For `interval_map(std::move(other))`: `other` will be constructed from `Map(Map&&)` if it exists. (it does)
* For `interval_map(other)`: `other` will be constructed from `Map(const Map&)` as `other` is an lvalue.

Reply:
> A constructor like `interval_set(Map other)` captures both copy and move:

... at the expense of an additional move in all cases except when constructing from a temporary map like that librbd test does.  I noted "maximum efficiency" in my comment for a reason ;)

Reply:
Ok, that's true.

Reply:
I checked it again, I'm afraid we must revert this commit. 
Because from the comment in https://github.com/ceph/ceph/blob/main/src/osd/PrimaryLogPG.cc#L5962, it seems the `m` could be modified by the `objects_readv_sync()`. If with `&&` we couldn't get the modified map, right ? @batrick @idryomov

Reply:
I think all that's required is to change `ReplicatedBackend::objects_readv_sync` back to taking an lvalue it can assign to, i.e. `map<uint64_t, uint64_t>& m`. The

https://github.com/ceph/ceph/commit/7c53a69eaeec46ff45a7f9105b6a2a85f50be125#diff-2ce7003fee423619cfc88a3c94727f5a329ef821bd65a84c4e57bf803e64dbeeR264

should be okay since we move the map back after success (and do not re-encode if it fails).

Reply:
Yeah, correct. I will partially revert the changes.